schema/validator.go

```diff
@@ -20,90 +20,121 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"os"
 	"regexp"
 
 	digest "github.com/opencontainers/go-digest"
 	v1 "github.com/opencontainers/image-spec/specs-go/v1"
-	"github.com/xeipuuv/gojsonschema"
+	"github.com/santhosh-tekuri/jsonschema/v5"
 )
 
-// Validator wraps a media type string identifier
-// and implements validation against a JSON schema.
+// Validator wraps a media type string identifier and implements validation against a JSON schema.
 type Validator string
 
-type validateFunc func(r io.Reader) error
-
-var mapValidate = map[Validator]validateFunc{
-	ValidatorMediaTypeImageConfig: validateConfig,
-	ValidatorMediaTypeDescriptor:  validateDescriptor,
-	ValidatorMediaTypeImageIndex:  validateIndex,
-	ValidatorMediaTypeManifest:    validateManifest,
-}
-
 // ValidationError contains all the errors that happened during validation.
+//
+// Deprecated: this is no longer used by [Validator].
 type ValidationError struct {
 	Errs []error
 }
 
+// Error returns the error message.
+//
+// Deprecated: this is no longer used by [Validator].
 func (e ValidationError) Error() string {
 	return fmt.Sprintf("%v", e.Errs)
 }
 
 // Validate validates the given reader against the schema of the wrapped media type.
 func (v Validator) Validate(src io.Reader) error {
-	buf, err := io.ReadAll(src)
-	if err != nil {
-		return fmt.Errorf("unable to read the document file: %w", err)
-	}
-
-	if f, ok := mapValidate[v]; ok {
-		if f == nil {
-			return fmt.Errorf("internal error: mapValidate[%q] is nil", v)
+	// run the media type specific validation
+	if fn, ok := validateByMediaType[v]; ok {
+		if fn == nil {
+			return fmt.Errorf("internal error: mapValidate is nil for %s", string(v))
 		}
-		err = f(bytes.NewReader(buf))
+		// buffer the src so the media type validation and the schema validation can both read it
```

Comment:
Is there some way we could buffer this closer to where it's used more than once so the need for buffering (and the potential eventual removal of the need for buffering) is more obvious?

Reply:
Resolved in our discussion during today's meeting (it's better than before, and we're a bit tired since the reader is part of the public interface).